numpy/ma/core.py

```diff
@@ -1285,25 +1285,48 @@ def __str__(self):
 ###############################################################################
 
 
-def _recursive_make_descr(datatype, newtype=bool_):
-    "Private function allowing recursion in make_descr."
+def _replace_dtype_fields_recursive(dtype, primitive_dtype):
+    "Private function allowing recursion in _replace_dtype_fields."
     # Do we have some name fields ?
-    if datatype.names:
+    if dtype.names:
         descr = []
-        for name in datatype.names:
-            field = datatype.fields[name]
+        for name in dtype.names:
+            field = dtype.fields[name]
             if len(field) == 3:
                 # Prepend the title to the name
                 name = (field[-1], name)
-            descr.append((name, _recursive_make_descr(field[0], newtype)))
-        return descr
+            descr.append((name, _replace_dtype_fields_recursive(field[0], primitive_dtype)))
+        new_dtype = np.dtype(descr)
+
     # Is this some kind of composite a la (np.float,2)
-    elif datatype.subdtype:
-        mdescr = list(datatype.subdtype)
-        mdescr[0] = _recursive_make_descr(datatype.subdtype[0], newtype)
-        return tuple(mdescr)
+    elif dtype.subdtype:
+        mdescr = list(dtype.subdtype)
+        mdescr[0] = _replace_dtype_fields_recursive(dtype.subdtype[0], primitive_dtype)
+        new_dtype = np.dtype(tuple(mdescr))
+
+    # this is a primitive type, so do a direct replacement
     else:
-        return newtype
+        new_dtype = primitive_dtype
+
+    # preserve identity of dtypes
+    if new_dtype == dtype:
+        new_dtype = dtype
+
+    return new_dtype
+
+
+def _replace_dtype_fields(dtype, primitive_dtype):
```

Comment:
Possible scope for public visibility? I can see "convert all fields to object/bool" being marginally desirable elsewhere